server/src/main/java/org/elasticsearch/index/mapper/BlockDocValuesReader.java

```diff
@@ -398,6 +398,12 @@ private static class SingletonDoubles extends BlockDocValuesReader {
 
         @Override
         public BlockLoader.Block read(BlockFactory factory, Docs docs, int offset, boolean nullsFiltered) throws IOException {
+            if (docValues instanceof BlockLoader.OptionalColumnAtATimeReader direct) {
+                BlockLoader.Block result = direct.tryReadDoubles(factory, docs, offset, toDouble);
```

Comment:
If we have an OverLong vector or block, I think we don't need to introduce tryReadDoubles. Instead, we can just read longs from the codec and perform the conversion here. This would significantly reduce the complexity in the codec. We can add a method to BlockLoaderFactory that converts a block to doubles with a ToDouble function.